content/Link Tracking with StimulusReflex - Part II.md

```diff
@@ -0,0 +1,613 @@
+
+![Header image](images/stimulus_reflex.jpeg)
+
+## Creating the model
+
+Let's start with some good old fashioned Rails code, we'll create the migration for our links and the associated model:
+
+
+```ruby
+class CreateShortenedLinks < ActiveRecord::Migration[6.0]
+  def change
+    create_table :shortened_links do |t|
+      t.string :name, null: false
+      t.string :shortened_path, null: false
+      t.string :original_url, null: false
+      t.integer :views_count, default: 0
+
+      t.timestamps
+    end
+  end
+end
+```
+
+```ruby
+# app/models/shortened_link
+
+class ShortenedLink < ApplicationRecord
+  validates :name, :shortened_path, :original_url, presence: true
+  validates :shortened_path, uniqueness: true
+end
+```
+
+Notice how we added some simple validations to make it more fun.
+
+## Listing links
+
+Let's start by the simplest of actions: listing the links. To do so, we'll create the following files:
+
+```ruby
+# app/controllers/shortened_links_controller.rb
+
+class ShortenedLinksController < ApplicationController
+  def index
+    @shortened_links ||= ShortenedLink.all
+  end
+end
+```
+
+```erb
+# app/views/shortened_links/index.html.erb
+
+<h1>Shortened Links</h1>
+
+<div class="row">
+  <div class="col-8">
+    <div class="row">
+      <div class="col"><b>Name</b></div>
+      <div class="col"><b>Shortened path</b></div>
+      <div class="col"><b>Original url</b></div>
+    </div>
+
+    <% @shortened_links.each do |shortened_link| %>
+      <div class="row">
+        <div class="col"><%= shortened_link.name %></div>
+        <div class="col"><%= shortened_link.shortened_path %></div>
+        <div class="col"><%= shortened_link.original_url %></div>
+      </div>
+    <% end %>
+  </div>
+</div>
+```
+
+We are also going to add [Bootstrap](https://getbootstrap.com/) styling so it doesn't look so dull, add the following tag to `app/views/layouts/application.html.erb`
+
+```html
+<link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.5.0/css/bootstrap.min.css">
+```
+
+## The link form
+
+What we did in the last section was pretty normal, huh? Let's keep going. Now we'll add a button to create new links by stimulating a Reflex. To do so, we add this markup and the corresponding StimulusReflex code:
+
+```erb
+# app/views/shortened_links/index.html.erb
+
+<h1>Shortened Links</h1>
+
+<div data-controller="shortened-links">
+  <button data-action="click->shortened-links#newLink" class="btn btn-primary">New Shortened Link</button>
+
+  <!-- links table ...-->
+</div>
+```
+
+```javascript
+// app/javascript/controllers/shortened_links_controller.js
+
+import { Controller } from "stimulus"
+import StimulusReflex from 'stimulus_reflex'
+
+export default class extends Controller {
+  connect () {
+    StimulusReflex.register(this)
+  }
+
+  newLink() {
+    this.stimulate('ShortenedLinksReflex#new')
+  }
+}
+```
+
+Now we should create that Reflex in `app/reflexes/shortened_links_reflex.rb`
```

Comment:
Is there a generator for this?

Reply:
Yeah but I'd rather keep things simple, generators on tutorials add too much "magic" for beginners at least IMO